File: Erlang/0003.erl
-export([length_of_longest_substring/1]).

-spec length_of_longest_substring(S :: unicode:unicode_binary()) -> integer().
length_of_longest_substring(S) ->
    length_of_longest_substring(S, 0, 0, 0, #{}).

-spec length_of_longest_substring(S :: unicode:unicode_binary(), Left :: integer(), Right :: integer(), MaxLength :: integer(), CharMap :: #{}) -> integer().
length_of_longest_substring(<<>>, _Left, _Right, MaxLength, _CharMap) ->
    MaxLength;
length_of_longest_substring(S, Left, Right, MaxLength, CharMap) ->
    <<CurrentChar:8, Rest/binary>> = S,
    case maps:get(CurrentChar, CharMap, -1) of
        LastIndex when LastIndex >= Left ->
            NewLeft = LastIndex + 1,
            NewCharMap = maps:put(CurrentChar, Right, CharMap),
            length_of_longest_substring(Rest, NewLeft, Right + 1, MaxLength, NewCharMap);
        _ ->
            NewCharMap = maps:put(CurrentChar, Right, CharMap),
            NewMaxLength = max(MaxLength, Right - Left + 1),
            length_of_longest_substring(Rest, Left, Right + 1, NewMaxLength, NewCharMap)
    end.

max(A, B) when A > B -> A;
max(_A, B) -> B.

File: Elixir/0003.exs
defmodule Solution do
  @spec length_of_longest_substring(s :: String.t()) :: integer
  def length_of_longest_substring(s) do
    {max_length, _, _} = 
      s
      |> String.graphemes()
      |> Enum.with_index()
      |> Enum.reduce({0, 0, %{}}, fn {char, right}, {max_length, left, char_map} ->
        left = 
          if Map.has_key?(char_map, char) && char_map[char] >= left do
            char_map[char] + 1
          else
            left
          end

        char_map = Map.put(char_map, char, right)

        max_length = max(max_length, right - left + 1)

        {max_length, left, char_map}
      end)

    max_length
  end
end
